### DISCUSSION TITLE
reactive text not updating properly in API context

### DISCUSSION THREAD

**Discussion by itsOdell** [Textualize/textual#4690] [Q&A]
Ive been trying to get reactive texts to update based on the response of a 3rd party Api since 2 days, but the text just never updates. Heres a simple example

```
Class Status(Static):
    current_status = reactive("initializing")

    def compose(self):
        yield Static(self.current_status)

   async def on_mount(self):
        await asyncio.sleep(1)
        res = await get("some_url")
        self.current_status = "getting results..."
        await asyncio.sleep(1)
        if res.ok:
           self.current_status = "you passed"
       else:
           self.current_status = "you failed"

Class MyApp(App):
    def compose(self):
        yield Header()
        yield Status()
        yield Footer()
```

**Comment by itsOdell** [comment]
the text never updates/rerenders, even when i force a rereder of the app

**Reply by willmcgugan** [reply]
The compose method will not be called again, until you recompose. https://textual.textualize.io/guide/reactivity/#recompose

You will also run in to issues if you sleep inside a message handler, which will prevent the UI from updating.

You might want to join the discord server if you have any more questions.

**Reply by itsOdell** [reply]
I have also tried with the render() method and force update using self.update() or self.query_one(name).value = "new value". Unfortutnely it all hangs and seems to be stuck

**Reply by itsOdell** [reply]
but i appreciate your help

**Reply by willmcgugan** [reply]
Nothing will update until you leave the on_mount event handler. Unless you run it concurrently, with a worker https://textual.textualize.io/guide/workers/

Here's how you update a static: https://textual.textualize.io/widgets/static/#textual.widgets._static.Static.update

I'd recommend spending some time reading through the docs. Particularly the tutorial. You should find it straightforward once it clicks.

**Comment by TomJGooding** [comment]
I've quickly modified your example which might help your understanding, but I'd recommend having another read of the docs about [reactvity](https://textual.textualize.io/guide/reactivity/). Also notice the use of a [worker](https://textual.textualize.io/guide/workers/) so request doesn't stop the UI from updating.

```python
import asyncio
import random

from textual import work
from textual.app import App, ComposeResult
from textual.reactive import reactive
from textual.widgets import Static


class Status(Static):
    current_status = reactive("initializing")

    def watch_current_status(self) -> None:
        self.update(self.current_status)

    def on_mount(self) -> None:
        self.get_result()

    @work
    async def get_result(self) -> None:
        await asyncio.sleep(1)  # simulate work
        self.current_status = "getting results..."
        await asyncio.sleep(1)  # simulate work
        passed = random.choice([True, False])
        if passed:
            self.current_status = "you passed"
        else:
            self.current_status = "you failed"


class ExampleApp(App):
    def compose(self) -> ComposeResult:
        yield Status()


if __name__ == "__main__":
    app = ExampleApp()
    app.run()
```
